stdlib/Random/src/normal.jl

```diff
@@ -116,6 +126,111 @@ end
 end
 
 
+## BigFloat
+
+struct MPZ_t # like BigInt, but immutable
+    alloc::Cint
+    size::Cint
+    d::Ptr{Limb}
+end
+
+mutable struct GMPRandState <: Sampler{BigFloat}
+    seed::MPZ_t          # where to store the RNG state
+    alg::Cint            # unused by MPFR, we store here the distribution
+    alg_data::Ptr{Cvoid} # GMPRandFnPtr
+end
+
+GMPRandState() = GMPRandState(MPZ_t(zero(Cint), zero(Cint), C_NULL), zero(Cint), C_NULL)
+
+
+mutable struct GMPRandFnPtr
+    randseed_fn  ::Ptr{Cvoid}
+    randget_fn   ::Ptr{Cvoid}
+    randclear_fn ::Ptr{Cvoid}
+    randiset_fn  ::Ptr{Cvoid}
+end
+
+# the three following functions should be unused
+_randseed_fn(::GMPRandState, ::BigInt)       = @assert false
+_randclear_fn(::GMPRandState)                = @assert false
+_randiset_fn(::GMPRandState, ::GMPRandState) = @assert false
+
+function _randget_fn(s::GMPRandState, dst::Ptr{Limb}, nbits::Culong)
+    rng = unsafe_pointer_to_objref(Ptr{Cvoid}(s.seed.d))
+    sz = (nbits + bits_in_Limb - 1) ÷ bits_in_Limb
+    rest = nbits % bits_in_Limb
+    a = UnsafeView(dst, Int(sz))
+    _randget_fn_fill!(rng, a) # rng is not inferred here, so let's have a function barrier
+    if rest != 0
+        a[end] &= ~(typemax(Limb) << rest)
+    end
+    nothing
+end
+
+@noinline _randget_fn_fill!(rng, a) = rand!(rng, a)
+
+const JuliaGMPRandFnPtr = GMPRandFnPtr(C_NULL, C_NULL, C_NULL, C_NULL)
+
+function get_JuliaGMPRandFnPtr()
+    if JuliaGMPRandFnPtr.randget_fn == C_NULL
+        JuliaGMPRandFnPtr.randseed_fn  = @cfunction(_randseed_fn,  Cvoid, (Ref{GMPRandState}, Ref{BigInt}))
+        JuliaGMPRandFnPtr.randget_fn   = @cfunction(_randget_fn,   Cvoid, (Ref{GMPRandState}, Ptr{Limb}, Culong))
+        JuliaGMPRandFnPtr.randclear_fn = @cfunction(_randclear_fn, Cvoid, (Ref{GMPRandState},))
+        JuliaGMPRandFnPtr.randiset_fn  = @cfunction(_randiset_fn,  Cvoid, (Ref{GMPRandState}, Ref{GMPRandState}))
+    end
+    pointer_from_objref(JuliaGMPRandFnPtr)
+end
+
+struct Normal{T<:AbstractFloat} end
+struct Exponential{T<:AbstractFloat} end
+
+function Sampler(::Type{<:AbstractRNG}, d::Union{Normal{BigFloat},Exponential{BigFloat}}, ::Repetition)
+    s = GMPRandState()
+    s.alg_data = get_JuliaGMPRandFnPtr()
+    s.alg = Cint(d isa Normal{BigFloat})
+    return s
+end
+
+function rand(rng::AbstractRNG, s::GMPRandState)
+    z = BigFloat()
+    if !ismutable(rng)
+        rng = MutableRNG(rng)
```

Comment:
replacing this line by `return rand(Mutable(rng), s)` (and fixing a method ambiguity) seems to solve the segfault problem, but I'd prefer some advice from someone who understands...